Ask the external provider for type info first in visitAnyNominalTypeRef

augusto2112 opened the pull request:
C++ imported types have Builtin type metadata. Ask the external provider before checking with builder.

This also has the added benefit
of not parsing all the typeref metadata in cases we don't have metadata for a type, but the external provider can resolve it.

augusto2112 commented:
@swift-ci smoke test

adrian-prantl commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -2226,6 +2226,18 @@ class LowerType
+    }
+
     auto FD = TC.getBuilder().getFieldTypeInfo(TR);
     if (FD == nullptr || FD->isStruct()) {
```
I wonder if it were safer to just move it into the `if (FD == nullptr || FD->isStruct())` condition?

kastiglione replied:
Would we want `ExternalTypeInfo` to be consulted when `FD->isStruct()` is true? In other words, should external lookup happen only when `FD == nullptr`

augusto2112 replied:
I think there are two situations where we want to ask the external provider:
- We don't have any metadata for the type, ask the external provider.
- We have metadata for the type and it's a builtin (builtins are structs), ask the external provider as it may have more accurate information.

kastiglione replied:
in those two situations, should they be checked after `TC.getBuilder().getFieldTypeInfo(TR)`?

kastiglione replied:
currently this change checks before finding out whether either of those two cases apply.

augusto2112 replied:
Yup! Just updated the patch :)

augusto2112 committed:
Ask the external provider for type info first in visitAnyNominalTypeRef

C/C++ imported types have Builtin type metadata. Ask the external
provider when the type info we have is builtin.

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5307

@swift-ci smoke test

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5307

@swift-ci smoke test macOS

adrian-prantl commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -2226,29 +2226,43 @@ class LowerType
   }
 
   const TypeInfo *visitAnyNominalTypeRef(const TypeRef *TR) {
+    auto AskExternalTypeInfoProvider = [&]() -> const TypeInfo * {
```
What do you think about `queryExternalTypeInfoProvider` ?

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5307

@swift-ci smoke test

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5307

@swift-ci smoke test macOS

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
Looks like TestSwiftFoundation crashed.

augusto2112 commented:
@adrian-prantl it's actually `TestSwiftFoundationTypeNotification.py` (you have to test this together with https://github.com/apple/llvm-project/pull/5307). I haven't been able to reproduce this failure locally yet.

augusto2112 commented:
https://github.com/apple/llvm-project/pull/5307

@swift-ci smoke test macOS
